bash_completion

```diff
@@ -789,23 +789,38 @@ _quote_readline_by_ref()
 
 # This function performs file and directory completion. It's better than
 # simply using 'compgen -f', because it honours spaces in filenames.
-# @param $1  If `-d', complete only on directories.  Otherwise filter/pick only
-#            completions with `.$1' and the uppercase version of it as file
-#            extension.
-#
+# @param $1 Complete filenames matching `.$1' and the uppercase version of it.
+#           Ignored with `-d`.
+# OPTIONS
+#   -C dir  Change to dir before generating completions
+#   -d      Complete only on directories
 _filedir()
```

Comment:
Problem here is that the `-o filenames` at end of the function won't take effect as we are no longer in the dir we changed to. So e.g. appending slashes to dir names does not happen. Not sure what to do about that.

(Commenting here because it seems the UI here won't let me comment that far away from changed lines.)

Reply:
Doing the `compopt -o filenames` inside each `compgen` subshell doesn't seem to make any difference.

Reply:
Hmm, this is a good point. But after another thinking of it, I guess this issue also existed in the other approach in #827, so I think this is not an issue specific to this approach.

If we are to solve this issue within this PR, maybe we need to manually suffix a slash in the shell code.

- One way is to test the filename in a `for` loop suffix / if it is a directory name.
- Another way is to use `compgen -S / -d` for directory names, and get non-directory filenames using another way (`find`, etc.).
- Or we may switch between the above two based on the number of files in the specified directory.

Maybe we don't have to too much care about it, but the opposite issue is that when a filename in the specified directory has the same name as a directory in the current directory, it would be unexpectedly suffixed by `/`.

Reply:
Yes, the problem existed in the other implementation in #827, I don't know why I didn't notice that. In my later tests I also noticed that it ends up changing the dir to /usr/share/mime in various scenarios, something I didn't notice originally either for some reason.

I'm going to try out yet another different approach for #827 before returning to this.

Reply:
Tweaked some, tests pass but there are still issues:
```
$ xx() { _filedir -C foo; }
$ mkdir -p foo/bar foo/quux
$ complete -F xx x
$ x <TAB>
bar/   quux/   # <-- this is all good
$ x ba<TAB>
bar/   quux/   # <-- this is not 
```

The problem of filename elsewhere matching a dir in current dir you mentioned also persists.